### PR TITLE
Turn off -Wcast-align inline for musl-libc

### PR DESCRIPTION
**Pull request by jasonwilliams200OK**
Tracking issue: #6312

In musl-libc, CMSG_NXTHDR typecasts char* to cmsghdr* which causes
clang to throw cast-align warning. This change suppresses the warning
inline.

Note, with `-Werror` and `-Weverything`, compiler treats this warning
as the error:

```ash
/root/corefx/src/Native/System.Native/pal_networking.cpp:1311:31: error: cast from 'unsigned char *' to 'struct cmsghdr *'
increases
      required alignment from 1 to 4 [-Werror,-Wcast-align]
             controlMessage = CMSG_NXTHDR(&header, controlMessage))
                              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/sys/socket.h:267:10: note: expanded from macro 'CMSG_NXTHDR'
        ((struct cmsghdr *)__CMSG_NEXT(cmsg))))
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

musl-libc: https://fossies.org/dox/musl-1.0.5/include_2sys_2socket_8h.html#a4f70853d9633fb80428189cd7bb292d0 (click on `__CMSG_NEXT` in that macro to see `unsigned char*` cast)

[clang/gcc]-libc: http://lxr.free-electrons.com/source/include/linux/socket.h#L89 (click on `__cmsg_nxthdr` in the called macro to notice that it returns `cmsghdr *`)

### CODE REVIEW

**Comment by jasonwilliams200OK** [comment]
Test Innerloop Ubuntu Debug Build and Test
Test Innerloop Windows_NT Release Build and Test

File: src/Native/System.Native/pal_networking.cpp [hunk 1]
```diff
@@ -1272,6 +1272,21 @@ static int32_t GetIPv6PacketInformation(cmsghdr* controlMessage, IPPacketInforma
     return 1;
 }
 
+cmsghdr* GET_CMSG_NXTHDR(msghdr* mhdr, cmsghdr* cmsg)
+{
+#ifndef _GLIBC_
```
Comment (stephentoub) [inline src/Native/System.Native/pal_networking.cpp]: Are you sure this is defined on, for example, our Ubuntu builds?

Comment (jasonwilliams200OK) [inline src/Native/System.Native/pal_networking.cpp]: There was an `_` missing like `__GLIBC__` I have just added it; which should only be defined for non-musl libc standard headers such as on all Ubuntus which ship with glibc, OS X and BSDs.

**Comment by stephentoub** [comment]
This is an unfortunate amount of ceremony to work around what would seem to be an issue in the ```musl-libc``` implementation itself rather than in our consumption of it.  Can it not be addressed there instead?  And is the warning pointing out a real issue that shouldn't be ignored? Are there other cases like this, or is this the final PR necessary to successfully compile with musl-libc?

cc: @pgavlin, @ericeil, @nguerrera

**Comment by pgavlin** [comment]
Seconding Steve's concerns: this does feel more like a `musl-libc` issue than a CoreFX issue. Given that we're applying a workaround, it would be nice to file an issue to track removing the workaround and reference the issue # in the code.

**Comment by nguerrera** [comment]
This looks a case where clang is unaware that a given header is "system". Normally, warnings are suppressed in system headers. I believe this is how you can communicate from CMake to Clang that more paths are "system":  https://github.com/dotnet/corefx/blob/master/src/Native/configure.cmake#L13

**Comment by nguerrera** [comment]
See https://github.com/dotnet/corefx/pull/3487 for last time we dealt with this.

**Comment by jasonwilliams200OK** [comment]
@stephentoub, yes this is the last patch for Alpine Linux. Current master without this patch breaks the native build at 70%. Going by their source trees, they have been shipping that macro with same behavior for quite some.

@pallavit, not sure where to file a bug, would really appreciate if @benpye or @ncopa can help with filing the issue to the concerned party.

@nguerrera, this was the first thing I tried, but unfortunately that didn't worked. I also tried couple of other methods to turn off this warning from cmake for non `__GLIBC__` systems, but no luck. It seems like expansion of macro is excluded from that rule where cmake differentiate between system and project headers.

**Comment by pgavlin** [comment]
@jasonwilliams200OK an issue on CoreFX is fine--AFAICT musl does not have an issue tracker besides their mailing list.

**Comment by jasonwilliams200OK** [comment]
@pgavlin, done => #6312

File: src/Native/System.Native/pal_networking.cpp [hunk 2]
```diff
@@ -1278,6 +1278,22 @@ static int32_t GetIPv6PacketInformation(cmsghdr* controlMessage, IPPacketInforma
     return 1;
 }
 
+cmsghdr* GET_CMSG_NXTHDR(msghdr* mhdr, cmsghdr* cmsg)
+{
+#ifndef __GLIBC__
+// Tracking issue: #6312
+// In musl-libc, CMSG_NCTHDR typecasts char* to cmsghdr* which causes
```
Comment (stephentoub) [inline src/Native/System.Native/pal_networking.cpp]: > CMSG_NCTHDR

CMSG_NXTHDR?

Comment (jasonwilliams200OK) [inline src/Native/System.Native/pal_networking.cpp]: Fixed.

**Pull request merged**